tools/database/src/resharding.rs

```diff
@@ -0,0 +1,188 @@
+
+            // Open write db.
+            let write_path = if write_path.is_absolute() {
+                PathBuf::from(&write_path)
+            } else {
+                home_dir.join(&write_path)
+            };
+            let write_path = write_path.as_path();
+            let write_config = &config.config.store;
+            let write_db =
+                RocksDB::open(write_path, write_config, Mode::ReadWrite, Temperature::Hot)?;
+            let write_db = Arc::new(write_db);
+
+            // Prepare the full mixed db.
+            // It will read, in order, from write, hot and cold.
+            // It will write only to the write db.
+            let mixed_db = MixedDB::new(split_db, write_db, ReadOrder::WriteDBFirst);
+
+            // The only way to create a Store is to go through NodeStorage.
+            let storage = NodeStorage::new(mixed_db);
+            storage.get_hot_store()
+        };
+        Ok(store)
+    }
+
+    fn get_chain(&self, mut config: NearConfig, home_dir: &Path) -> Result<Chain, anyhow::Error> {
+        let store = self.get_store(home_dir, &mut config)?;
+
+        let epoch_manager = EpochManager::new_arc_handle(store.clone(), &config.genesis.config);
+        let genesis_epoch_config = epoch_manager.get_epoch_config(&EpochId::default())?;
+        initialize_sharded_genesis_state(
+            store.clone(),
+            &config.genesis,
+            &genesis_epoch_config,
+            Some(home_dir),
+        );
+        let shard_tracker = ShardTracker::new(
+            TrackedConfig::from_config(&config.client_config),
+            epoch_manager.clone(),
+        );
+        let runtime_adapter =
+            NightshadeRuntime::from_config(home_dir, store, &config, epoch_manager.clone())?;
+        let chain_genesis = ChainGenesis::new(&config.genesis.config);
+        let client_config = config.client_config;
+        let chain_config = ChainConfig {
+            save_trie_changes: client_config.save_trie_changes,
+            background_migration_threads: client_config.client_background_migration_threads,
+            resharding_config: client_config.resharding_config,
+        };
+        let chain = Chain::new(
+            Clock::real(),
+            epoch_manager,
+            shard_tracker,
+            runtime_adapter,
+            &chain_genesis,
+            DoomslugThresholdMode::TwoThirds,
+            chain_config,
+            None,
+            Arc::new(RayonAsyncComputationSpawner),
+            MutableConfigValue::new(None, "validator_signer"),
+        )
+        .unwrap();
+        Ok(chain)
+    }
+
+    // Rely on the regular config but make sure it's configured correctly for
+    // the on demand resharding. It's executed while the node is not running so
+    // it should be as fast as possible - there should be no throttling.
+    fn check_resharding_config(config: &mut NearConfig) {
```

Comment:
JFYI you can play with the batch size a bit and see if increasing it would speed up the process.